File: scratch/test_proxy.go
package main

import (
	"io"
	"log"
	"net"
	"net/http"
)

func handleTunnel(w http.ResponseWriter, r *http.Request) {
	destConn, err := net.Dial("tcp", r.Host)
	if err != nil {
		http.Error(w, err.Error(), http.StatusServiceUnavailable)
		return
	}
	w.WriteHeader(http.StatusOK)
	hijacker, ok := w.(http.Hijacker)
	if !ok {
		http.Error(w, "Hijacking not supported", http.StatusInternalServerError)
		return
	}
	clientConn, _, err := hijacker.Hijack()
	if err != nil {
		http.Error(w, err.Error(), http.StatusServiceUnavailable)
		return
	}
	go func() {
		defer destConn.Close()
		defer clientConn.Close()
		io.Copy(destConn, clientConn)
	}()
	go func() {
		defer destConn.Close()
		defer clientConn.Close()
		io.Copy(clientConn, destConn)
	}()
}

func handleHTTP(w http.ResponseWriter, r *http.Request) {
	r.RequestURI = ""
	resp, err := http.DefaultTransport.RoundTrip(r)
	if err != nil {
		http.Error(w, err.Error(), http.StatusServiceUnavailable)
		return
	}
	defer resp.Body.Close()
	for k, vv := range resp.Header {
		for _, v := range vv {
			w.Header().Add(k, v)
		}
	}
	w.WriteHeader(resp.StatusCode)
	io.Copy(w, resp.Body)
}

func main() {
	handler := http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		log.Printf("[Proxy LOG] Received: %s %s %s", r.RemoteAddr, r.Method, r.URL.String())
		if r.Method == http.MethodConnect {
			handleTunnel(w, r)
		} else {
			handleHTTP(w, r)
		}
	})
	log.Println("Local test proxy server listening on port 8888...")
	log.Fatal(http.ListenAndServe(":8888", handler))
}

File: scratch/test_proxy.js
const http = require('http');
const net = require('net');
const url = require('url');

const server = http.createServer((req, res) => {
  console.log(`[Upstream Proxy] ${req.method} ${req.url}`);
  const parsedUrl = url.parse(req.url);
  const options = {
    hostname: parsedUrl.hostname,
    port: parsedUrl.port || 80,
    path: parsedUrl.path,
    method: req.method,
    headers: req.headers
  };

  const proxyReq = http.request(options, (proxyRes) => {
    res.writeHead(proxyRes.statusCode, proxyRes.headers);
    proxyRes.pipe(res);
  });

  proxyReq.on('error', (err) => {
    console.error(`[Upstream HTTP Error] ${err.message}`);
    res.writeHead(502);
    res.end();
  });

  req.pipe(proxyReq);
});

server.on('connect', (req, clientSocket, head) => {
  console.log(`[Upstream Proxy] CONNECT ${req.url}`);
  const [host, port] = req.url.split(':');
  
  const serverSocket = net.connect(port || 443, host, () => {
    clientSocket.write('HTTP/1.1 200 Connection Established\r\n\r\n');
    serverSocket.write(head);
    serverSocket.pipe(clientSocket);
    clientSocket.pipe(serverSocket);
  });

  serverSocket.on('error', (err) => {
    console.error(`[Upstream HTTPS Error] ${err.message}`);
    clientSocket.end('HTTP/1.1 502 Bad Gateway\r\n\r\n');
  });

  clientSocket.on('error', () => {
    serverSocket.end();
  });
});

const PORT = 8888;
server.listen(PORT, '0.0.0.0', () => {
  console.log(`Local upstream proxy server listening on port ${PORT}...`);
});
